trie/sync.go

```diff
@@ -193,10 +193,22 @@ func (s *Sync) AddCodeEntry(hash common.Hash, path []byte, parent common.Hash) {
 	s.schedule(req)
 }
 
-// Missing retrieves the known missing nodes from the trie for retrieval.
-func (s *Sync) Missing(max int) []common.Hash {
-	var requests []common.Hash
-	for !s.queue.Empty() && (max == 0 || len(requests) < max) {
+// Missing retrieves the known missing nodes from the trie for retrieval. To aid
+// both eth/6x style fast sync and snap/1x style state sync, the paths of trie
+// nodes are returned too, as well as separate hash list for codes.
+//
+// The returned paths are {account path, storage path} tuples, the second half
+// missing in case of an account trie node. The last component of the tuple is
+// always compact-encoded. If the tuple contains two items, the account path is
+// keybyte-encoded since we know it's always 32 bytes so no need for the compact
+// flag byte.
```

Comment:
I'm wondering if this whole encoding scheme could be made simpler by using some new types instead of raw byte slices..? Don't have a suggestion though, just seems rather complicated as is

Reply:
Deleted this part and made it into a new type with some more extended documentation.